text/###-v2-provider-interface.md

```diff
@@ -0,0 +1,402 @@
+
+This also starts the process of decoupling IQX's API from the provider interface
+because removing marshmallow means we don't get the 2 transformations it performs
+`numpy array -> list` and `complex numbers -> [re, im]`. Basically when v1.5
+starts merging we are saying that the `qobj.to_dict()`'s output is not the same
+as `json.loads()` from what the ibmq provider sends on the wire.
+
+After v1.5 is complete we can work on adding a new providers v2 interface
+that basically treats the providers interface as a python interface instead of
+as a pseudo-json api. The basic idea is that we use terra's objects as the
+what is passed back and forth to providers instead of the intermediate formats
+that exist today.
+
+For example, the new interface for running circuits (or pulse schedules) will
+look like:
+
+```python
+qc = QuantumCircuit()
+provider = my_provider()
+backend = my_provider.get_backend()
+backend.set_config(shots=2048)
+backend.run([qc]*100)
+```
+The abstract classes defining this new interface will live in:
+`qiskit.providers.v2` in the terra repo and for terra's functionality
+that interacts with anything as part of the provider interface (eg transpiler
+passes that use backend configuration) will need to check if it's using a
+provider based on v1 or v2 prior to interacting with it.
+
+A base backend abstract class will be defined as:
+
+```python
+from abc import ABC
+
+class BaseBackend(ABC):
+    def __init__(self):
+        configuration = self._default_config()
+
+    @classmethod
+    @abstractmethod
+    def _default_config(cls):
+        pass
+
+    def set_configuration(self, **fields):
+        configuration.update_config(**fields)
+
+    @property
+    def configuration(self):
+        return self.configuration
+
+    @abstractmethod
+    def run(self, circuits):
+        pass
+```
+
+With a configuration class that looks like:
+
+```python
+
+from abc import ABC
+
+
+class Configuration(ABC):
+
+    @property
+    @abstractmethod
+    def num_qubits(self):
+        pass
+
+    @property
+    @abstractmethod
+    def name(self):
+        pass
+```
+
+The required fields for the base configuration are kept minimal to maximize
+compatibility with all providers. Each provider is expected to define it's
+own configuration subclasses which will have any extra required fields that
+are specific to that backend.
+
+On the other side of the interface, interacting with a running circuit and
+it's results a new job class will be defined. This will merge in the standard
+results functionality (it's still present).
+
+```python
+from abc import ABC
+
+class BaseJob(ABC):
+    """A base class for representing an async job on a backend."""
+
+    job_id = None
+    metadata = None
+
+    def __init__(self, job_id, backend):
+
+    def backend(self):
+        return
+
+    @abstractmethod
+    def result():
+        pass
+
+    @abstractmethod
+    def status():
+
+    def cancel():
+        raise NotImplementedError
```

Comment:
What is the benefit of raising `NotImplementedError` here? If it's an optional method, why not just leave it out of `BaseJob` (so it doesn't show up in autocomplete/docs)? If it's required, isn't `@abstractmethod` a better choice?

Reply:
This is to make it optional. I left it it because we want this interface to be standardized between providers that implement this feature. We don't want to leave the interface freeform but don't want to force every provider to implement it. Were you thinking of defining this somewhere else?

Reply:
My main concern is that it puts the burden on the users to know whether their `job` instance supports `cancel()` (or to except `NotImplementedError`), especially since the `cancel()` method will show up in autocomplete/docs.